### PR TITLE
virtio-fs: New case set capability

### PR DESCRIPTION
**Pull request by zhenyzha**
New case set capability and then run virtiofsd

ID: 1912231,1934910
Signed-off-by: Zhenyu Zhang <zhenyzha@redhat.com>

### CODE REVIEW

**Comment by zhenyzha** [comment]
(1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.arm64-pci: PASS (77.57 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.auto.arm64-pci: PASS (77.97 s)

@YongxueHong @menli820 pls help me to review the patch if you are free, thanks.

**Comment by menli820** [comment]
@xiagao  pls help to review it， thanks

**Comment by zhenyzha** [comment]
```
 (1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.arm64-pci: PASS (74.11 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.arm64-pci: PASS (79.79 s)
```

The update test step is complete.
@YongxueHong @xiagao pls help me to review the patch if you are free, thanks.

**Comment by zhenyzha** [comment]
@xiagao pls help to review it， thanks

**Comment by xiagao** [comment]
Sorry for update late,  I thought I have submitted my comments already, but just now I found it failed as network issue.
BTW, could you update step7-8 for RHEL-196267(polarion case)

File: qemu/tests/cfg/virtio_fs_set_capability.cfg [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                - cap_dac_read_search:
+                    only virtio_fs_set_capability..with_cache.auto
+                    capability = cap_dac_read_search
+                - cap_sys_admin:
+                    only virtio_fs_set_capability..with_cache.none
+                    capability = cap_sys_admin
+                    fs_binary_extra_options += ' -o sandbox=chroot -o xattr -o xattrmap=':map::user.virtiofsd.:''
+                    cmd_yum_attr = 'yum install -y attr'
+                    cmd_set_trusted = 'setfattr -n trusted.test ${fs_dest}'
+                    cmd_get_trusted = 'getfattr -m '' ${fs_dest}'
```
[the remark this answers is not in the document]
Comment (zhenyzha) [reply qemu/tests/cfg/virtio_fs_set_capability.cfg:75]: But polarion case step5 has windows test steps, do we need to update polarion case to linux only?

File: qemu/tests/virtio_fs_set_capability.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+        source = os.path.join(base_dir, source)
+
+    sock_path = os.path.join(data_dir.get_tmp_dir(),
+                             '-'.join(('avocado-vt-vm1', 'viofs', 'virtiofsd.sock')))
+
+    # set capability
+    cmd_capsh_drop = (cmd_capsh_drop % capability)
+    prompt = params.get("shell_prompt", "[#$]")
+    error_context.context("Remove capability on host.", logging.info)
+    session = aexpect.ShellSession(cmd_capsh_drop, auto_close=False,
```
[the remark this answers is not in the document]
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py:110]: Same as above, if don't need to test this case on windows, this script will be updated to` linux only` .

**Comment by zhenyzha** [comment]
> Sorry for update late, I thought I have submitted my comments already, but just now I found it failed as network issue.
> BTW, could you update step7-8 for RHEL-196267(polarion case)

The current code already contains step7-8 for RHEL-196267 (polarion case)

File: qemu/tests/virtio_fs_set_capability.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/virtio_fs_set_capability.py:110]: Need to test windows guest, but only test share file basic function in this scenario(see polarion steps,they are updated)

File: qemu/tests/cfg/virtio_fs_set_capability.cfg [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            cmd_md5 = 'md5sum %s'
+            io_timeout = 120
+            Windows:
+                cmd_dd = 'dd if=/dev/random of=%s bs=1M count=200'
+                cmd_md5 = "%s: && md5sum.exe %s"
+            cmd_capsh_print = 'capsh --print'
+            cmd_capsh_drop = 'capsh --drop=%s --'
+            variants:
+                - cap_dac_read_search:
+                    only virtio_fs_set_capability..with_cache.auto
```
Comment (xiagao) [inline qemu/tests/cfg/virtio_fs_set_capability.cfg:67]: the two scenarios have the specified cache, how about adding fs_binary_extra_options in each variant.
If later we have another property or need to test all cache,then add them

File: qemu/tests/cfg/virtio_fs_set_capability.cfg (hunk 1 above)
[the remark this answers is not in the document]
Comment (xiagao) [reply qemu/tests/cfg/virtio_fs_set_capability.cfg:75]: Also need to test win guest under without cap_sys_admin ,just do basic function.

File: qemu/tests/virtio_fs_set_capability.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+            output = session.cmd_output(cmd_get_trusted)
+            logging.info("The guest file trusted is %s.", output)
+            if 'trusted.test' not in output:
+                test.fail('Set trusted failed.')
+
+            process.run(cmd_yum_attr)
+            output = str(process.run('getfattr %s' % fs_source).stdout.strip())
+            logging.info("The host file trusted is %s.", output)
+            if host_attributes not in output:
+                test.fail('Set trusted failed.')
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:267]: Could you set the fail info with "It's failed to check the trusted info from host."

File: qemu/tests/virtio_fs_set_capability.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+            session.cmd(cmd_create_file)
+            session.cmd(cmd_set_capability)
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:270]: add logging "Set captability for guest file."

File: qemu/tests/virtio_fs_set_capability.py [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            output = session.cmd_output(cmd_get_capability)
+            logging.info("The guest file capability is %s.", output)
+            if 'cap_net_raw+eip' not in output:
+                test.fail('Set capability failed.')
+            session.cmd(cmd_echo_file)
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:275]: pls add logging info "modify file content and check the file capability"

File: qemu/tests/virtio_fs_set_capability.py (hunk 5 above)
Comment (xiagao) [reply qemu/tests/virtio_fs_set_capability.py:270]: I mean add steps.

File: qemu/tests/cfg/virtio_fs_set_capability.cfg (hunk 3 above)
Comment (zhenyzha) [reply qemu/tests/cfg/virtio_fs_set_capability.cfg:67]: As you can see from line 46, the requirements for future expansion are taken into account here. If you want to add `fs_binary_extra_options` in the future, just remove only.

File: qemu/tests/cfg/virtio_fs_set_capability.cfg (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhenyzha) [reply qemu/tests/cfg/virtio_fs_set_capability.cfg:75]: Update later

File: qemu/tests/virtio_fs_set_capability.py (hunk 4 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py:267]: Okay, update later.

File: qemu/tests/virtio_fs_set_capability.py (hunk 5 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py:270]: Okay, update later.

File: qemu/tests/virtio_fs_set_capability.py (hunk 6 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py:275]: Okay, update later.

**Comment by zhenyzha** [comment]
```
 (1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.arm64-pci: PASS (74.06 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.arm64-pci: PASS (78.93 s)

```
The update test step is complete.
@xiagao @YongxueHong  pls help me to review the patch if you are free, thanks.

**Comment by xiagao** [comment]
@zhenyzha   git am the new patch, but hit some error. 
(1/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (114.56 s)
 (2/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: ERROR: Command 'yum install -y attr' failed.\nstdout: b'Updating Subscription Management repositories.\nUnable to read consumer identity\n\nThis system is not registered to Red Hat Subscription Management. You can use subscription-manager to register.\n\nbeaker-A... (77.96 s)
 (3/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: ERROR: Timeout expired while waiting for shell command to complete: 'capsh --print'    (output: "[root@hp-dl385g10-19 kar]# [root@hp-dl385g10-19 kar]# Current: = cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linu... (62.92 s)
 (4/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: ERROR: Timeout expired while waiting for shell command to complete: 'capsh --print'    (output: "[root@hp-dl385g10-19 kar]# [root@hp-dl385g10-19 kar]# Current: = cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,... (62.91 s)

**Comment by zhenyzha** [comment]
> @zhenyzha git am the new patch, but hit some error.
> (1/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (114.56 s)
> (2/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: ERROR: Command 'yum install -y attr' failed.\nstdout: b'Updating Subscription Management repositories.\nUnable to read consumer identity\n\nThis system is not registered to Red Hat Subscription Management. You can use subscription-manager to register.\n\nbeaker-A... (77.96 s)
> (3/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: ERROR: Timeout expired while waiting for shell command to complete: 'capsh --print' (output: "[root@hp-dl385g10-19 kar]# [root@hp-dl385g10-19 kar]# Current: = cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linu... (62.92 s)
> (4/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: ERROR: Timeout expired while waiting for shell command to complete: 'capsh --print' (output: "[root@hp-dl385g10-19 kar]# [root@hp-dl385g10-19 kar]# Current: = cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,... (62.91 s)

It looks like your host is not equipped with a yum warehouse, If it is a CI triggered job, it will help you with yum. 
`'yum install -y attr' failed.\nstdout: b'Updating Subscription Management repositories.\nUnable to read consumer identity\n\nThis system is not registered to Red Hat Subscription Management.`
I don't have windows env, could you provide an env for me to debug?

**Comment by zhenyzha** [comment]
@luckyh  Could you pls help review this patch,
The Windows guest device fails to be inserted, but the Linux guest can be inserted normally.
Detailed log I put in bug-1912231.
I tried to use `prepare_hotplug()` before inserting. But the issue was not solved.
Thanks a lot in advance.

```
 (1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (106.13 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: ERROR: viofs driver is not running (422.60 s)

```

**Comment by zhenyzha** [comment]
```
 (1/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (176.69 s)
 (2/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: PASS (130.70 s)
 (3/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (164.99 s)
 (4/4) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: PASS (162.47 s)
```
@xiagao @YongxueHong pls help me to review the patch if you are free, thanks.

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Comment by xiagao** [comment]
@zhenyzha Hit yum install attr error info.
 (1/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1229.51 s)
 (2/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.rh_kernel_update.q35: PASS (115.95 s)
 (3/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (100.43 s)
 (4/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: ERROR: Command 'yum install -y attr' failed.\nstdout: b'Updating Subscription Management repositories.\nUnable to read consumer identity\n\nThis system is not registered to Red Hat Subscription Management. You can use subscription-manager to register.\n\nbeaker-t... (51.27 s)

**Comment by zhenyzha** [comment]
> el850

@xiagao hello It seems that the repo in rhel.8.5.0 is not ready yet, this is not an issue with this code.

**Comment by xiagao** [comment]
> > el850
> 
> @xiagao hello It seems that the repo in rhel.8.5.0 is not ready yet, this is not an issue with this code.

But I think we'd better resolve it in automation test. @YongxueHong Do you have any idea about this situation?

**Comment by YongxueHong** [comment]
> > > el850
> > 
> > 
> > @xiagao hello It seems that the repo in rhel.8.5.0 is not ready yet, this is not an issue with this code.
> 
> But I think we'd better resolve it in automation test. @YongxueHong Do you have any idea about this situation?

I think we could disable 8.5.0 in this patch with `no rhel.8.5.0` in cfg, enable it when the repo is ready.
How about? @zhenyzha  @xiagao

**Comment by zhenyzha** [comment]
> > > > the latest version
> > > 
> > > 
> > > @xiagao hello It seems that the repo in the latest version is not ready yet, this is not an issue with this code.
> > 
> > 
> > But I think we'd better resolve it in automation test. @YongxueHong Do you have any idea about this situation?
> 
> I think we could disable the latest version in this patch with `no the latest version` in cfg, enable it when the repo is ready.
> How about? @zhenyzha @xiagao

@xiagao @YongxueHong 
First of all, we had better not disclose the latest version of the information.
Secondly, this latest version is only in the planning stage and has not yet begun to prepare for testing. So its repo is not complete, I think it is understandable. If you try some merged patches and need to use yum cmd, it will also fail. So this is not a code issue, but a test strategy issue.
In the product planning stage, we discuss whether a repo is complete or not, which is not in my scope of responsibility.
Finally, as I said, this is a test strategy issue. If it is necessary to make it clear that there is no test latest version at present, I agree with yhong's suggestion. It can be added in cfg.

**Comment by xiagao** [comment]
> > > > > the latest version
> > > > 
> > > > 
> > > > @xiagao hello It seems that the repo in the latest version is not ready yet, this is not an issue with this code.
> > > 
> > > 
> > > But I think we'd better resolve it in automation test. @YongxueHong Do you have any idea about this situation?
> > 
> > 
> > I think we could disable the latest version in this patch with `no the latest version` in cfg, enable it when the repo is ready.
> > How about? @zhenyzha @xiagao
> 
> @xiagao @YongxueHong
> First of all, we had better not disclose the latest version of the information.
> Secondly, this latest version is only in the planning stage and has not yet begun to prepare for testing. So its repo is not complete, I think it is understandable. If you try some merged patches and need to use yum cmd, it will also fail. So this is not a code issue, but a test strategy issue.
> In the product planning stage, we discuss whether a repo is complete or not, which is not in my scope of responsibility.
> Finally, as I said, this is a test strategy issue. If it is necessary to make it clear that there is no test latest version at present, I agree with yhong's suggestion. It can be added in cfg.

@zhenyzha @YongxueHong
Debug more,actually the package is ready in rhel850.
The command is run in both guest and host. no matter with rhel840 or 850,attr can be installed. Meanwhile, it always failed on my host as some repos are not available,such as beaker-task.After deleting it,everything go well.
So, there is no problem now.

File: qemu/tests/virtio_fs_set_capability.py [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+                    test.fail("It's failed to check the capability of the guest.")
+                logging.info("Modify file content and check the file capability.")
+                session.cmd(cmd_echo_file)
+                output = session.cmd_output(cmd_get_capability)
+                logging.info("The guest change file capability is %s.", output)
+                if file_capability in output:
+                    test.fail('Capability result doesn‘t meet expectations.')
+
+        if cmd_dd:
+            logging.info("Creating file under %s inside guest.", fs_dest)
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:282]: suggest use error_context.context() to print steps in debug log.

File: qemu/tests/virtio_fs_set_capability.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        fs_dest = "%s:" % volume_letter
+
+    guest_file = os.path.join(fs_dest, 'fs_test')
+    logging.info("The guest file in shared dir is %s.", guest_file)
+
+    try:
+        # No extended attributes (file steams) in virtio-fs for windows
+        if not is_windows:
+            if cmd_set_trusted:
+                host_attributes = params["host_attributes"]
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:250]: Add step to test  trusted attribute inside guest .
e.g.
error_context.context("Trusted attribute test without CAP_SYS_ADMIN  for linux guest", logging.info)

File: qemu/tests/virtio_fs_set_capability.py [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+                guest_trusted = params["guest_trusted"]
+                file_capability = params["file_capability"]
+                error_context.context("Set a trusted on guest.", logging.info)
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:253]: change to logging.info()

File: qemu/tests/virtio_fs_set_capability.py [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+                session.cmd(cmd_yum_attr)
+                session.cmd(cmd_set_trusted)
+                output = session.cmd_output(cmd_get_trusted)
+                logging.info("The guest file trusted is %s.", output)
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:257]: change logging content to "Failed to check the  trusted attribute from guest, the output is %s"

File: qemu/tests/virtio_fs_set_capability.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+                if guest_trusted not in output:
+                    test.fail("It's failed to check the trusted info from the guest.")
+
+                process.run(cmd_yum_attr)
+                output = str(process.run('getfattr %s' % fs_source).stdout.strip())
+                logging.info("The host file trusted is %s.", output)
+                if host_attributes not in output:
+                    test.fail("It's failed to check the trusted info from the host.")
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:265]: change logging content to "Failed to check the  trusted attribute from host, the output is %s"

File: qemu/tests/virtio_fs_set_capability.py [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+
+                session.cmd(cmd_create_file)
+                logging.info("Set capability for guest files.")
+                session.cmd(cmd_set_capability)
+                output = session.cmd_output(cmd_get_capability)
+                logging.info("The guest file capability is %s.", output)
+                if file_capability not in output:
+                    test.fail("It's failed to check the capability of the guest.")
+                logging.info("Modify file content and check the file capability.")
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:274]: Change it to a step.

File: qemu/tests/virtio_fs_set_capability.py [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+                session.cmd(cmd_echo_file)
+                output = session.cmd_output(cmd_get_capability)
+                logging.info("The guest change file capability is %s.", output)
+                if file_capability in output:
+                    test.fail('Capability result doesn‘t meet expectations.')
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:279]: test.fail('Still can get capability after file content is changed.')

File: qemu/tests/virtio_fs_set_capability.py [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+                    test.fail("It's failed to check the trusted info from the guest.")
+
+                process.run(cmd_yum_attr)
+                output = str(process.run('getfattr %s' % fs_source).stdout.strip())
+                logging.info("The host file trusted is %s.", output)
+                if host_attributes not in output:
+                    test.fail("It's failed to check the trusted info from the host.")
+
+                session.cmd(cmd_create_file)
+                logging.info("Set capability for guest files.")
```
Comment (xiagao) [inline qemu/tests/virtio_fs_set_capability.py:268]: Add step to test trusted attribute inside guest .
e.g.
error_context.context("Privileged capabilities test without CAP_SYS_ADMIN for linux guest", logging.info)

File: qemu/tests/virtio_fs_set_capability.py (hunk 7 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py]: Corrected

File: qemu/tests/virtio_fs_set_capability.py (hunk 9 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py]: Corrected

File: qemu/tests/virtio_fs_set_capability.py (hunk 10 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py]: Corrected

File: qemu/tests/virtio_fs_set_capability.py (hunk 11 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py]: Corrected

File: qemu/tests/virtio_fs_set_capability.py (hunk 12 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py:280]: Corrected

File: qemu/tests/virtio_fs_set_capability.py (hunk 13 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py]: Corrected

File: qemu/tests/virtio_fs_set_capability.py (hunk 14 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py]: Corrected

File: qemu/tests/virtio_fs_set_capability.py (hunk 8 above)
Comment (zhenyzha) [reply qemu/tests/virtio_fs_set_capability.py:252]: Corrected

**Comment by zhenyzha** [comment]
@xiagao pls help to review it， thanks

**Comment by xiagao** [comment]
ACK
 (1/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (108.54 s)
 (2/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: PASS (128.04 s)
 (3/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_dac_read_search.with_cache.auto.q35: PASS (161.49 s)
 (4/4) Host_RHEL.m8.u5.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.virtio_fs_set_capability.remove_capability.cap_sys_admin.with_cache.none.q35: PASS (161.98 s)

**Review by xiagao — APPROVED** [review]

**Pull request merged**